llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/StatsRecordingThreadPool.java

```diff
@@ -161,98 +149,44 @@ private void setupMDCFromNDC(final Callable<V> actualCallable) {
-              bytesWrittenDelta = threadFSStats.getBytesWritten();
-              readOpsDelta = threadFSStats.getReadOps();
-              largeReadOpsDelta = threadFSStats.getLargeReadOps();
-              writeOpsDelta = threadFSStats.getWriteOps();
-            }
-            tezCounters.findCounter(scheme, FileSystemCounter.BYTES_READ)
-                .increment(bytesReadDelta);
-            tezCounters.findCounter(scheme, FileSystemCounter.BYTES_WRITTEN)
-                .increment(bytesWrittenDelta);
-            tezCounters.findCounter(scheme, FileSystemCounter.READ_OPS).increment(readOpsDelta);
-            tezCounters.findCounter(scheme, FileSystemCounter.LARGE_READ_OPS)
-                .increment(largeReadOpsDelta);
-            tezCounters.findCounter(scheme, FileSystemCounter.WRITE_OPS)
-                .increment(writeOpsDelta);
+          LlapThreadLocalStatistics currentStats = new LlapThreadLocalStatistics(mxBean);
+          currentStats.subtract(statsBefore).fill(tezCounters);
 
-            if (LOG.isDebugEnabled()) {
-              LOG.debug("Updated stats: instance: {} thread name: {} thread id: {} scheme: {} " +
-                      "bytesRead: {} bytesWritten: {} readOps: {} largeReadOps: {} writeOps: {}",
-                  actualCallable.getClass().getSimpleName(), thread.getName(), thread.getId(),
-                  scheme, bytesReadDelta, bytesWrittenDelta, readOpsDelta, largeReadOpsDelta,
-                  writeOpsDelta);
-            }
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Updated stats: instance: {} thread name: {} thread id: {} stats: {}",
+                actualCallable.getClass().getSimpleName(), thread.getName(), thread.getId(), currentStats);
           }
         } else {
           LOG.warn("File system statistics snapshot before execution of thread is null." +
-                  "Thread name: {} id: {} allStats: {}", thread.getName(), thread.getId(),
-              statsBefore);
+                  "Thread name: {} id: {}", thread.getName(), thread.getId());
         }
       } else {
         LOG.warn("TezCounters is null for callable type: {}",
             actualCallable.getClass().getSimpleName());
       }
     }
   }
+
+  private static <V> TezCounters getTezCounters(Callable<V> actualCallable) {
+    TezCounters tezCounters = null;
+    // add tez counters for task execution and llap io
+    if (actualCallable instanceof TaskRunner2Callable) {
+      TaskRunner2Callable taskRunner2Callable = (TaskRunner2Callable) actualCallable;
+      // counters for task execution side
+      tezCounters = taskRunner2Callable.addAndGetTezCounter(FileSystemCounter.class.getName());
+    } else if (actualCallable instanceof TezCounterSource) {
+      // Other counter sources (currently used in LLAP IO).
+      tezCounters = ((TezCounterSource) actualCallable).getTezCounters();
+    } else {
+      LOG.warn("Unexpected callable {}; cannot get counters", actualCallable);
```

Comment:
Instead of warning, shouldn't an exception thrown?

Reply:
no, this is no expected, but should not turn the task fail

Reply:
Ok.